### ISSUE TITLE
Renovate picks wrong / default(?) repo path for shared config if configured with encrypted npmToken and npmrc

### ISSUE THREAD

**Issue by maplesteve** [renovatebot/config-help#557]
**What Renovate type are you using?**

GitHub App

**Describe the bug**

We have a shared config in a private Github repo. Other repos reference this and Renovate does its excellent job. Since we host packages in the private GitHub Package registry, we added a PAT (encrypted as advisd in the docs) as the encrypted `npmToken` and the GitHub registry as `npmrc` properties to the `renovate.json` file in repos which use those private packages.

With that extended config, Renovate does not load the config from the private repo anymore but tries to load a config from a different (assumed) repo: `renovate-config-config` which does not exist - because, well our shared config repo is named different...

This leads to a `Preset fetch error` and in the end to the error message 
```
Repository has invalid config
{
  "error": {
    "validationError": "Cannot find preset's package (config:base). Note: this is a *nested* preset so please contact the preset author if you are unable to fix it yourself."
  }
}
```

**Did you see anything helpful in debug logs?**

The "migrated configs" in both examples are correct; they name the correct `extends` value.

Fail in Job-ID:  #147773735
```
DEBUG: Dependency lookup failure: unauthorized
{
  "pkgUrl": "https://npm.pkg.github.com/ourOrgName/renovate-config-config",
   ...
```
Here, IMO `unauthorized` is the normal response from GitHub which is also raised, when a (private) repo does not exist.

Working example: Job-ID #147793369; same `extends` in `renovate.json` as in above example; but only this property.

**To Reproduce**

Working `renovate.json`:
```
{
    "extends": [
        "github> ourOrgName/bar-renovate-presets"
    ]
}
```

Failing `renovate.json`:
```
{
    "extends": ["github>ourOrgName/bar-renovate-presets"],
    "encrypted": {
        "npmToken": "foo=="
    },
    "npmrc": "//npm.pkg.github.com/:_authToken=${NPM_TOKEN}\nregistry=https://npm.pkg.github.com/ourOrgName"
}
```

**Additional context**

**Comment by maplesteve** [comment]
@viceice 
The shared config is not published as a package, but as a repository; it's not loaded from a registry.

The repo with the working `renovate.json` has the same `.npmrc` file as the failing one. The `npmrc` config in the failing `renovate.json` is the also the same (besides the token variable which is `NPM_TOKEN` there). So, both are pointing to the same registry.

**Comment by rarkins** [comment]
This line in your `.npmrc` tells Renovate to use GitHub for all packages:

```
registry=https://npm.pkg.github.com/ourOrgName
```

If Renovate does exactly that, then it's behaving per design/per configuration and there is no mistake. 

If it's *not* using that registry when it's defined within a preset, it could be because the `.npmrc` value is only getting loaded *after* Renovate has already checked the normal presets. Neither is a mistake and @viceice is right - you should be using scoped package names.

**Comment by maplesteve** [comment]
@rarkins thanks. I don't get it. Do you mean, that the shared config **has** to be provided and referenced as a package instead of a repo, if the `renovate.json` contains the `npmrc` property? 

The [documentation](https://docs.renovatebot.com/config-presets/#preset-hosting) advises to use the repo approach: "An additional benefit of using source code hosting is that the same token/authentication can be reused by Renovate in case you want to make your config private."

> If it's not using that registry when it's defined within a preset, it could be because the .npmrc value is only getting loaded after Renovate has already checked the normal presets.

Since the `extends` points to a repo, not to a package, why does Renovate try to load a package?

**Comment by rarkins** [comment]
Your `npmrc` is wrong because the `registry=` line is too general. The failure happens as soon as Renovate attempts to retrieve an npm package not hosted on GitHub. I’m guessing you reference at least one in your custom preset.

**Comment by maplesteve** [comment]
Yes, the preset references `config:base` and `:semanticCommitTypeAll(chore)` where `config:base` is loaded from the package `renovate-config-config` (right?).

> The failure happens as soon as Renovate attempts to retrieve an npm package not hosted on GitHub.

GitHub Packages registry routes/caches requests to npm automatically. With exactly the same `npmrc` I can `npm i renovate-config-config` without a problem. It's fetched from/through the GitHub registry.

**Comment by rarkins** [comment]
Can the problem be reproduced using purely public repos? If we uncover a bug in Renovate then we can move this issue back to the main repo

**Comment by rarkins** [comment]
If only can be reproduced using private then that’s a useful data point too, just makes it a little harder to troubleshoot

**Comment by maplesteve** [comment]
I created a shared config in a public repo and referenced it in `renovate.json` in the private repo.
With `npmrc` in `renovate.json`, it fails (job id #148002695).
Without `npmrc` it works (job id #148002786).

So, same behaviour. No difference, if the shared config is private or public.

**Comment by rarkins** [comment]
Can you describe how you generate the GitHub Packages token?

**Comment by rarkins** [comment]
One more thing : please try running the Renovate CLI against this same repo using a personal access token for auth. Does it fail when the npmrc is defined in renovate.json?

**Comment by maplesteve** [comment]
> Can you describe how you generate the GitHub Packages token?

It's a "normal" Personal Access Token with the scopes "read:packages, repo" which is crypted via https://renovatebot.com/encrypt

**Comment by maplesteve** [comment]
I installed the Renovate CLI, but didn't get this to work in any way. This is a different behaviour from running the Renovate GitHub App where the `renovate.json` with just the "extends" property works and Renovate loads the config source code from the private repo.

To make sure, the basic private access works, I issued the command `RENOVATE_TOKEN=*** npm i @orgName/private-package` with the following `.npmrc`:
```
//npm.pkg.github.com/:_authToken=${RENOVATE_TOKEN}
registry=https://npm.pkg.github.com/orgName
```
This works; package gets installed from GitHub registry.

Whatever combination I tried with the CLI, does not work. With/without `.npmrc` file, with/without `npmToken` and/or `npmrc` properties in `renovate.json`. I also tried passing the token via ENV variables and directly (uncrypted) in `npmToken`.

Further observation:
I set "--log-level trace" but the first log line is "INFO: Throwing preset error" which is IMO thrown [here](https://github.com/renovatebot/renovate/blob/16962a989f8f9627da6d600830f2535af3f4505b/lib/config/presets/index.ts#L193). But the traces from L151, L166 or L105 are not printed. If a non-existing RENOVATE_CONFIG_FILE is passed, a lot of info is printed (and a default config is loaded; no error, that the passed config does not exist).

**Comment by rarkins** [comment]
@maplesteve if you run with `LOG_LEVEL=trace` in env then the trace logging will kick in much earlier. When it's run via CLI it only starts once the CLI and other config parsing is complete, whereas LOG_LEVEL in env gets checked during logger initialization.

**Comment by maplesteve** [comment]
@rarkins The CLI fails while it tries to fetch the shared config; I think it doesnt' do the auth with the GitHub API correctly (or at all).
So the CLI approach won't help us here. Do you have any other idea how to make this work with the Renovate GitHub App?

**Comment by rarkins** [comment]
You would need a PAT with access to both repos. With reproduction I cant fix it. I actually think we might have an incompatibility with GitHub Packages but that would be independent of App vs CLI

**Comment by maplesteve** [comment]
In the CLI tests is no private package (GitHub registry) involved. Just the shared config, which is not a package but a private repo. The PAT has access to all repos (and the registry).

**Comment by maplesteve** [comment]
So, we're basically back to square one. As soon as the encrypted npmToken and the npmrc are part of the renovate.json file, the GitHub App it tries to load "renovate-config-config" from the wrong path and fails. As far as I can see from the logs, this happens even before the shared config is loaded (Job-ID #148531502).

> I actually think we might have an incompatibility with GitHub Packages

Yes, this seems to be the conclusion. Means, that Renovate can't be used to "watch" for updated packages which are in the private GitHub Packages registry for the time being.

**Comment by rarkins** [comment]
So what’s the simplest reproduction that can be tested against to verify that? 
- no presets 
- package.json with one outdated package
- npmrc with PAT pointing to GitHub packages?

**Comment by viceice** [comment]
@rarkins I thin i have an idea whats wrong. setting up a test case to check.

**Comment by viceice** [comment]
https://github.com/viceice/renovate-gh-npm-test/pull/1

**Comment by rarkins** [comment]
Possibly the app is applying its app token to the lookup, which doesn’t have permissions currently for packages

**Comment by viceice** [comment]
I've got it working now: https://github.com/viceice/renovate-gh-npm-test/pull/1

**Comment by rarkins** [comment]
Was the earlier config a valid/desired one?

**Comment by viceice** [comment]
*conclusion*
we need different `npmrc` settings for each manager.


*renovate*
```ini
//npm.pkg.github.com/viceice/:_authToken=XXXX
registry=https://npm.pkg.github.com/viceice
```

*yarn*
```ini
//npm.pkg.github.com/viceice/:_authToken=XXXX
registry=https://npm.pkg.github.com/viceice
always-auth=true
```

*npm*
```ini
//npm.pkg.github.com/:_authToken=XXXX
registry=https://npm.pkg.github.com/viceice
```

**Comment by viceice** [comment]
> Was the earlier config a valid/desired one?

yes, but he couldn't fetch preset after validation

**Comment by rarkins** [comment]
Is it just yarn that’s different? And why? They normally strive for compatibility

**Comment by viceice** [comment]
https://github.com/viceice/renovate-gh-npm-test/pull/4 now the npm auth problem is visible

**Comment by viceice** [comment]
`npm` and `yarn` can fallback to `_authToken=XXXX` but renovate always require `//npm.pkg.github.com/viceice/:_authToken=XXXX`
